Extract profile spherical velocity and data load

vivi235711 opened the pull request:
Enhance GAMER_ExtractProfile with reading hybrid scheme output, improved accuracy, shell-average velocity, and spherical velocity output.

vivi235711 committed:
Enhance 'GAMER_ExtractProfile' to read hybrid scheme data

vivi235711 committed:
Improve the differential accuracy in GAMER_ExtractProfile using Richardson extrapolation into second order

vivi235711 committed:
Remove the center of mass velocity for evaluating the shell-average of velocity

vivi235711 committed:
Add option 'O' to 'GAMER_ExtractProfile' for generating spherical velocity output

vivi235711 committed:
Enable differential accuracy as an option in GAMER_ExtractProfile

vivi235711 committed:
Correct CMV calculation, changed sin&cos to SIN&COS, and skip density = 0 region

vivi235711 committed:
Remove CMV using phase shift

vivi235711 committed:
Change to wave function derivative to calculate internal velocity; Output volume and mass in gamer_extract_profile

vivi235711 committed:
Correct velocity calculation

vivi235711 committed:
Remove center of mass velocity as an option 'C'

vivi235711 committed:
Fix error of unsupported model

vivi235711 committed:
Merge branch 'psidm' of https://github.com/hyschive/gamer-fork into extract_profile_sph_vel_and_data_load

vivi235711 committed:
Merge branch 'main' of https://github.com/gamer-project/gamer into extract_profile_sph_vel_and_data_load

hsinhaoHHuang reviewed (changes requested):
@vivi235711 
Thank you for providing this useful tool.
I have finished the PR review.
Please see my comments when you have time and let me know if they are not clear.
Thanks!

vivi235711 committed:
Style: Align code

vivi235711 committed:
Fix: HDF5 loading with ELBDM_Scheme

vivi235711 committed:
Refactor code based on @hsinhaoHHuang's comment

Rename variables, remove useless line, modify some if statements, and
add usage

vivi235711 committed:
Support: OutputSphVel in Hydro

vivi235711 commented on tool/analysis/gamer_extract_profile/GAMER_Functions/LoadData_HDF5.cpp:
```diff
@@ -160,6 +167,7 @@ void LoadData_HDF5( const char *FileName )
 #  elif ( MODEL == ELBDM )
    LoadField( "ELBDM_Mass",           &ELBDM_Mass,           H5_SetID_InputPara, H5_TypeID_InputPara, Fatal,  NullPtr,        -1, NonFatal );
    LoadField( "ELBDM_PlanckConst",    &ELBDM_PlanckConst,    H5_SetID_InputPara, H5_TypeID_InputPara, Fatal,  NullPtr,        -1, NonFatal );
+   LoadField( "ELBDMScheme",          &ELBDM_Scheme,         H5_SetID_Makefile,  H5_TypeID_Makefile, NonFatal, NullPtr,        -1, NonFatal );
```
I've set default `ELBDMScheme` = 1.

vivi235711 commented on tool/analysis/gamer_extract_profile/Main.cpp:
```diff
@@ -238,6 +242,330 @@ void GetMaxRho()
+#  warning : WAIT MHD !!!
+
+#  elif ( MODEL == ELBDM )
+   const int  NGhost = (ELBDM_GetVir) ? 2 : 0;
```
You are right. I've also removed the if `ELBDM_GetVir` judgement.

vivi235711 commented on tool/analysis/gamer_extract_profile/Main.cpp:
```diff
@@ -238,6 +242,330 @@ void GetMaxRho()
+                     exit( 1 );
+                  }
+
+#                 if   ( MODEL == HYDRO )
+//                evaluate the values on the shell
+#                 warning : WAIT HYDRO !!!
+
+#                 elif ( MODEL == MHD )
+#                 warning : WAIT MHD !!!
+
+#                 elif ( MODEL == ELBDM )
```
Ok. I moved for loop outside.

vivi235711 commented on tool/analysis/gamer_extract_profile/Main.cpp:
```diff
@@ -238,6 +242,330 @@ void GetMaxRho()
+#  error : ERROR : unsupported MODEL !!
+#  endif // MODEL
+
+// get the average values
+   for (int n=0; n<NShell; n++)    average_dens[n] /= volume[n];
+
+// get the average velocity
+#  if   ( MODEL == HYDRO )
+#  warning : WAIT HYDRO !!!
+
+#  elif ( MODEL == MHD )
+#  warning : WAIT MHD !!!
+
+#  elif ( MODEL == ELBDM )
+   if ( ELBDM_GetVir )
+   {
+      for (int n=0; n<NShell; n++)
+         if (nCount[n]!=0)  massAll += volume[n]*average_dens[n];
```
Thanks! I only left `massAll` and delete `nCount`, `volume` and `average_dens`.

vivi235711 commented on tool/analysis/gamer_extract_profile/Main.cpp:
```diff
@@ -238,6 +242,330 @@ void GetMaxRho()
+   if ( average_dens  != NULL )  delete [] average_dens;
+   cout << " Bulk CoM velocity = ("<< CMV[0]<< ", "<< CMV[1] << ", "<<CMV[2]<< ")" << endl;
+
+   cout << "Remove the motion of center-of-mass by phase shift" << endl;
```
Yeah. I'm still thinking about let `Remove_CMVel()` can support all model. Do you think only get the center of mass velocity here is better? Then we can subtract the velocity when we calculate the shell average and RMS. Please let me know if you have any thought.

vivi235711 commented on tool/analysis/gamer_extract_profile/Main.cpp:
```diff
@@ -410,26 +745,64 @@ void GetRMS()
                   ParDens = Field[p][PAR_DENS][k][j][i];
 
+
                   if ( ELBDM_GetVir )
```
I think `OutputSphere` is only needed when `ELBDM_GetVir` is enabled, but `ELBDM_GetVir` can function without `OutputSphere`. Therefore, I added an if statement at the calculation part and ensured that `OutputSphVel `is automatically set to true whenever `ELBDM_GetVir` is enabled.

vivi235711 commented on tool/analysis/gamer_extract_profile/Main.cpp:
```diff
@@ -478,6 +860,14 @@ void GetRMS()
                   RMS[ShellID][Var] += dv*pow( double(wt_abs )-Average[ShellID][Var], 2.0 );  Var++; }
 
+                  if ( OutputSphere ) {
+                  RMS[ShellID][Var] += dv*Dens*pow( double(v_sph[0])-Average[ShellID][Var], 2.0 );  Var++;
+                  RMS[ShellID][Var] += dv*Dens*pow( double(v_sph[1])-Average[ShellID][Var], 2.0 );  Var++;
+                  RMS[ShellID][Var] += dv*Dens*pow( double(v_sph[2])-Average[ShellID][Var], 2.0 );  Var++;
+                  RMS[ShellID][Var] += dv*Dens*pow( double(w_sph[0])-Average[ShellID][Var], 2.0 );  Var++;
+                  RMS[ShellID][Var] += dv*Dens*pow( double(w_sph[1])-Average[ShellID][Var], 2.0 );  Var++;
+                  RMS[ShellID][Var] += dv*Dens*pow( double(w_sph[2])-Average[ShellID][Var], 2.0 );  Var++; }
```
In my case, I want to get the kinetic energy, so the velocity is weighted by mass. I guess `vr` and `wr` want to get the fluid dynamic properties so using volume weighted? Anyway, I didn't change the old code and just add new variables here.
@hyschive May I ask for your thoughts?

hyschive replied:
I prefer weighting velocity by mass rather than volume.

vivi235711 commented on tool/analysis/gamer_extract_profile/Main.cpp:
```diff
@@ -716,24 +1116,61 @@ void ShellAverage()
 
                   if ( ELBDM_GetVir )
                   {
```
I would add this in new pr.

vivi235711 commented on tool/analysis/gamer_extract_profile/Main.cpp:
```diff
@@ -1229,6 +1705,15 @@ void Output_ShellAve()
                         sprintf( FileName[Var++], "%s", "AveWr-A"  );
                         sprintf( FileName[Var++], "%s", "AveWt-A"  );
    }
+   if ( OutputSphere )
+   {
+                        sprintf( FileName[Var++], "%s", "AveVr"    );
+                        sprintf( FileName[Var++], "%s", "AveVtheta");
+                        sprintf( FileName[Var++], "%s", "AveVphi"  );
+                        sprintf( FileName[Var++], "%s", "AveWr"    );
+                        sprintf( FileName[Var++], "%s", "AveWtheta");
+                        sprintf( FileName[Var++], "%s", "AveWphi"  );
+   }
```
Now can support hydro.
May I ask what hydro test problem you use to test this code?

vivi235711 commented:
@hsinhaoHHuang Thanks for your detailed reviewed! I've updated the code based on your comments. Please take a look when you have time.

vivi235711 committed:
Change RMS velocity to  mass-weighted

vivi235711 committed:
Turn on ELBDM_GetVir when -O is on in ELBDM

Based on @koarakawaii 's comment

hyschive commented:
@vivi235711 Please update this PR to synchronize it with the latest `main`.

@hsinhaoHHuang Please take another look at this PR (if needed) and approve it if you think it's ready to be merged.

hsinhaoHHuang commented on tool/analysis/gamer_extract_profile/Main.cpp:
```diff
@@ -1316,7 +1815,7 @@ void Output_ShellAve()
 
       if ( OutputAccMass )
-      fprintf( File, "  %13s  %13s", "AccMass", "AveDens" );
+      fprintf( File, "  %13s  %13s  %13s  %13s %13s","ShellMass", "ShellVol", "AccMass", "AccVolume", "AveDens" );
```
```suggestion
      fprintf( File, "  %13s  %13s  %13s  %13s  %13s", "ShellMass", "ShellVol", "AccMass", "AccVolume", "AveDens" );
```

hsinhaoHHuang commented on tool/analysis/gamer_extract_profile/Main.cpp:
[Style] There are some trailing spaces in this file (found by `grep -r -n ' $' Main.cpp`).

hsinhaoHHuang reviewed (changes requested):
Apologies for my late response. I have reviewed it again and left some comments. Please let me know your thoughts about them when you have time. Thank you!

hyschive commented:
@vivi235711
- The compilation fails for `MODEL=HYDRO`
- Please synchronize with the latest `main` branch.

koarakawaii commented:
When running the code with `MODEL = ELBDM` with `-V` option on, the RMS for `AveDens_*` is different when `-O` option is on/off (probably related to [this line](https://github.com/vivi235711/gamer/blob/extract_profile_sph_vel_and_data_load/tool/analysis/gamer_extract_profile/Main.cpp#L858), which should not touch the density data when `-O` option is off, but seems to still touch it because `NOut` seems to be `12` in this case).
